src/OpenSpaceToolkit/Astrodynamics/Solvers/LeastSquaresSolver.cpp

```diff
@@ -0,0 +1,419 @@
+{
+    return rmsUpdateThreshold_;
+}
+
+LeastSquaresSolver::Analysis LeastSquaresSolver::solve(
+    const State& anInitialGuessState,
+    const Array<State>& aReferenceStateArray,
+    const std::function<Array<State>(const State&, const Array<Instant>&)>& aGenerateStatesCallback,
+    const std::unordered_map<CoordinateSubset, VectorXd>& anInitialGuessSigmas,
+    const std::unordered_map<CoordinateSubset, VectorXd>& aReferenceStateSigmas
+) const
+{
+    if (!anInitialGuessState.isDefined())
+    {
+        throw ostk::core::error::runtime::Undefined("Initial guess state");
+    }
+
+    if (aReferenceStateArray.isEmpty())
+    {
+        throw ostk::core::error::runtime::Undefined("Reference states");
+    }
+
+    // Setup state builders
+    const StateBuilder estimationStateBuilder(anInitialGuessState);
+    const Shared<const Frame>& estimationStateFrame = anInitialGuessState.accessFrame();
+    const StateBuilder referenceStateBuilder(aReferenceStateArray[0]);
+    const Array<Shared<const CoordinateSubset>> referenceStateSubsets = referenceStateBuilder.getCoordinateSubsets();
+
+    // Get dimensions
+    const Size stateDimension = anInitialGuessState.accessCoordinates().size();
+    const Size referenceStateDimension = aReferenceStateArray[0].accessCoordinates().size();
+    const Size referenceStateCount = aReferenceStateArray.size();
+
+    // Validate dimensions
+    if (referenceStateDimension * referenceStateCount <= stateDimension)
+    {
+        throw ostk::core::error::RuntimeError(
+            "Reference State count * reference State dimension should be greater than estimated State dimension."
+        );
+    }
+
+    // Setup a priori and covariance matrices
+    VectorXd xApriori = VectorXd::Zero(stateDimension);
+    MatrixXd PAprioriInv;
+
+    if (!anInitialGuessSigmas.empty())
+    {
+        const VectorXd sigmas = expandSigmas(anInitialGuessSigmas, estimationStateBuilder);
+        PAprioriInv = sigmas.array().square().inverse().matrix().asDiagonal();
+    }
+    else
+    {
+        PAprioriInv = MatrixXd::Zero(stateDimension, stateDimension);
+    }
+
+    // Setup measurement covariance matrix
+    MatrixXd RInv;
+    if (!aReferenceStateSigmas.empty())
+    {
+        const VectorXd sigmas = expandSigmas(aReferenceStateSigmas, referenceStateBuilder);
+        RInv = sigmas.array().square().inverse().matrix().asDiagonal();
+    }
+    else
+    {
+        RInv = MatrixXd::Identity(referenceStateDimension, referenceStateDimension);
+    }
+
+    // Initialize state vectors
+    VectorXd XNom = anInitialGuessState.accessCoordinates();
+    VectorXd xHat = VectorXd::Zero(stateDimension);
+
+    // Get reference state coordinates
+    MatrixXd referenceStatesCoordinates(referenceStateCount, referenceStateDimension);
+    for (Size i = 0; i < referenceStateCount; ++i)
+    {
+        referenceStatesCoordinates.row(i) = aReferenceStateArray[i].inFrame(estimationStateFrame).accessCoordinates();
```

Comment:
I know it's the same, but just to be in case of future refactors:

```suggestion
        referenceStatesCoordinates.row(i) = aReferenceStateArray[i].inFrame(estimationStateFrame).extractCoordinates(referenceStateSubsets);
```

Reply:
Hmmmm this feels like overkill as the state builder and coordinate subsets are created from the reference state array...